### PR TITLE
[AMDGPU][Attributor] Make `AAAMDWavesPerEU` honor existing attribute

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114438?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#114438** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114438?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#114357** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/114357?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/CodeGen/AMDGPU/propagate-waves-per-eu.ll [hunk 1]
```diff
@@ -404,21 +404,20 @@ attributes #19 = { "amdgpu-waves-per-eu"="8,9" }
...
-; CHECK: attributes #[[ATTR15]] = { "amdgpu-no-agpr" "amdgpu-no-completion-action" "amdgpu-no-default-queue" "amdgpu-no-dispatch-id" "amdgpu-no-dispatch-ptr" "amdgpu-no-heap-ptr" "amdgpu-no-hostcall-ptr" "amdgpu-no-implicitarg-ptr" "amdgpu-no-lds-kernel-id" "amdgpu-no-multigrid-sync-arg" "amdgpu-no-queue-ptr" "amdgpu-no-workgroup-id-x" "amdgpu-no-workgroup-id-y" "amdgpu-no-workgroup-id-z" "amdgpu-no-workitem-id-x" "amdgpu-no-workitem-id-y" "amdgpu-no-workitem-id-z" "amdgpu-waves-per-eu"="6,9" "uniform-work-group-size"="false" }
+; CHECK: attributes #[[ATTR11]] = { "amdgpu-flat-work-group-size"="1,64" "amdgpu-no-agpr" "amdgpu-no-completion-action" "amdgpu-no-default-queue" "amdgpu-no-dispatch-id" "amdgpu-no-dispatch-ptr" "amdgpu-no-heap-ptr" "amdgpu-no-hostcall-ptr" "amdgpu-no-implicitarg-ptr" "amdgpu-no-lds-kernel-id" "amdgpu-no-multigrid-sync-arg" "amdgpu-no-queue-ptr" "amdgpu-no-workgroup-id-x" "amdgpu-no-workgroup-id-y" "amdgpu-no-workgroup-id-z" "amdgpu-no-workitem-id-x" "amdgpu-no-workitem-id-y" "amdgpu-no-workitem-id-z" "amdgpu-waves-per-eu"="1,123" "uniform-work-group-size"="false" }
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/propagate-waves-per-eu.ll:413]: Hmm, this is definitely wrong. Need to check.

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/propagate-waves-per-eu.ll:413]: Oh, this is not wrong. It is from users.

File: llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp [hunk 2]
```diff
@@ -215,6 +215,19 @@ class AMDGPUInformationCache : public InformationCache {
     return ST.getWavesPerEU(F, FlatWorkGroupSize);
   }
 
+  std::optional<std::pair<unsigned, unsigned>>
+  getWavesPerEUAttr(const Function &F) {
+    Attribute Attr = F.getFnAttribute("amdgpu-waves-per-eu");
+    if (!Attr.isStringAttribute())
+      return std::nullopt;
+    auto Val = parseRangeAttribute(Attr.getValueAsString());
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:223]: I don't see where this is defined

Comment (shiltian) [reply llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:223]: In the parent PR

**Comment by choikwa** [comment]
LGTM

**Review by choikwa — APPROVED** [review]

**Comment by shiltian** [comment]
### Merge activity

* **Dec 11, 4:38 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/114438).

**Pull request merged**